File: backend/app/schemas/social.py
"""Pydantic schemas for social OAuth and accounts."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class SocialAccountRead(BaseModel):
    """Social account response."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    platform: str
    created_at: datetime


class ConnectResponse(BaseModel):
    """OAuth connect response - redirect URL."""

    auth_url: str


class CallbackSuccess(BaseModel):
    """OAuth callback success."""

    message: str = "Аккаунт успешно подключён"
    account_id: UUID


class SocialAccountResponse(BaseModel):
    """Single social account in list."""

    id: UUID
    platform: str
    channel_id: str | None = None
    channel_title: str | None = None
    created_at: str | None


class SocialAccountsListResponse(BaseModel):
    """List of social accounts."""

    accounts: list[SocialAccountResponse]


class SocialAccountUpdate(BaseModel):
    """Update social account (display name)."""

    channel_title: str | None = Field(
        None, max_length=256, description="Отображаемое название канала"
    )


# OAuth App Credentials schemas


class OAuthAppCreate(BaseModel):
    """Create OAuth app credentials."""

    platform: str = Field(..., pattern="^(youtube|vk)$", description="Platform: youtube, vk")
    name: str = Field(..., min_length=1, max_length=256, description="Display name")
    client_id: str = Field(..., min_length=1, max_length=512)
    client_secret: str = Field(..., min_length=1, description="Will be encrypted")
    redirect_uri: str | None = Field(None, max_length=512)


class OAuthAppUpdate(BaseModel):
    """Update OAuth app credentials (partial)."""

    name: str | None = Field(None, min_length=1, max_length=256)
    client_id: str | None = Field(None, min_length=1, max_length=512)
    client_secret: str | None = Field(None, min_length=1, description="Will be encrypted")
    redirect_uri: str | None = Field(None, max_length=512)


class OAuthAppRead(BaseModel):
    """OAuth app credentials response (without client_secret)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID | None
    platform: str
    name: str
    client_id_masked: str = Field(..., description="Masked client_id (last 4 chars)")
    redirect_uri: str | None
    created_at: datetime
    updated_at: datetime


class OAuthAppListResponse(BaseModel):
    """List of OAuth app credentials."""

    apps: list[OAuthAppRead]

File: frontend/src/services/social.ts
import { api } from "./api";

export interface SocialAccount {
  id: string;
  platform: string;
  channel_id: string | null;
  channel_title: string | null;
  created_at: string | null;
}

export interface SocialAccountsListResponse {
  accounts: SocialAccount[];
}

export interface OAuthApp {
  id: string;
  user_id: string | null;
  platform: string;
  name: string;
  client_id_masked: string;
  redirect_uri: string | null;
  created_at: string;
  updated_at: string;
}

export interface OAuthAppListResponse {
  apps: OAuthApp[];
}

export interface OAuthAppCreate {
  platform: string;
  name: string;
  client_id: string;
  client_secret: string;
  redirect_uri?: string;
}

export interface OAuthAppUpdate {
  name?: string;
  client_id?: string;
  client_secret?: string;
  redirect_uri?: string | null;
}

export const socialService = {
  async getAccounts(): Promise<SocialAccount[]> {
    const response = await api.get<SocialAccountsListResponse>("/social/accounts");
    const accounts = response.data?.accounts ?? [];
    return Array.isArray(accounts)
      ? accounts.filter((a) => a && typeof a.id === "string" && a.id.length > 0)
      : [];
  },

  async connectPlatform(platform: string, oauthAppId: string): Promise<string> {
    const response = await api.get<{ auth_url: string }>(`/social/connect/${platform}?oauth_app_id=${oauthAppId}`);
    return response.data.auth_url;
  },

  async updateAccount(accountId: string, data: { channel_title?: string | null }): Promise<SocialAccount> {
    const response = await api.patch<SocialAccount>(`/social/accounts/${accountId}`, data);
    return response.data;
  },

  async disconnectAccount(accountId: string): Promise<void> {
    await api.delete(`/social/accounts/${accountId}`);
  },

  async syncVkProfile(accountId: string): Promise<SocialAccount> {
    const response = await api.post<SocialAccount>(`/social/accounts/${accountId}/sync-profile`);
    return response.data;
  },

  // OAuth Apps
  async getOAuthApps(platform?: string): Promise<OAuthApp[]> {
    const url = platform ? `/social/oauth-apps?platform=${platform}` : "/social/oauth-apps";
    const response = await api.get<OAuthAppListResponse>(url);
    return response.data.apps;
  },

  async createOAuthApp(data: OAuthAppCreate): Promise<OAuthApp> {
    const response = await api.post<OAuthApp>("/social/oauth-apps", data);
    return response.data;
  },

  async updateOAuthApp(appId: string, data: OAuthAppUpdate): Promise<OAuthApp> {
    const response = await api.patch<OAuthApp>(`/social/oauth-apps/${appId}`, data);
    return response.data;
  },

  async deleteOAuthApp(appId: string): Promise<void> {
    await api.delete(`/social/oauth-apps/${appId}`);
  },
};
